cmd/loadgenerator/main.go

```diff
@@ -0,0 +1,222 @@
+
+func loadMnemonic(mnemonic string) crypto.Seed {
+	seedbytes, err := passphrase.MnemonicToKey(mnemonic)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Cannot recover key seed from mnemonic: %v\n", err)
+		os.Exit(1)
+	}
+
+	var seed crypto.Seed
+	copy(seed[:], seedbytes)
+	return seed
+}
+
+func main() {
+	var cfg config
+	var err error
+	if cfg, err = loadConfig(); err != nil {
+		fmt.Fprintf(os.Stderr, "unable to load config : %v\n", err)
+		os.Exit(1)
+	}
+	fmt.Printf("Configuration file loaded successfully.\n")
+
+	seed := loadMnemonic(cfg.AccountMnemonic)
+	privateKey := crypto.GenerateSignatureSecrets(seed)
+	publicKey := basics.Address(privateKey.SignatureVerifier)
+
+	fmt.Printf("Spending account public key : %v\n", publicKey.String())
+
+	err = spendLoop(cfg, privateKey, publicKey)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "spend loop error : %v\n", err)
+		os.Exit(1)
+	}
+
+	os.Exit(0)
+}
+
+func isSpendRound(cfg config, round uint64) bool {
+	return cfg.RoundModulator == 0 || ((round+cfg.RoundOffset)%cfg.RoundModulator == 0)
+}
+
+func nextSpendRound(cfg config, round uint64) uint64 {
+	if cfg.RoundModulator == 0 {
+		return round
+	}
+	return ((round+cfg.RoundOffset)/cfg.RoundModulator)*cfg.RoundModulator + cfg.RoundModulator
+}
+
+func spendLoop(cfg config, privateKey *crypto.SignatureSecrets, publicKey basics.Address) (err error) {
+	restClient := client.MakeRestClient(*cfg.ClientURL, cfg.APIToken)
+
+	for {
+		waitForSpendingRound(restClient, cfg)
+		queueFull := generateTransactions(restClient, cfg, privateKey, publicKey)
+		if queueFull {
+			waitForNonSpendingRound(restClient, cfg)
+		}
+	}
+}
+
+func waitForNonSpendingRound(restClient client.RestClient, cfg config) {
+	var nodeStatus generatedV2.NodeStatusResponse
+	var err error
+	for {
+		nodeStatus, err = restClient.Status()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "unable to check status : %v", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		if !isSpendRound(cfg, nodeStatus.LastRound) {
+			return
+		}
+	waitForNextBlock:
+		// wait for the next round.
+		nodeStatus, err = restClient.WaitForBlock(basics.Round(nodeStatus.LastRound))
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "unable to wait for next round node status : %v", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		if !isSpendRound(cfg, nodeStatus.LastRound) {
+			return
+		}
+		goto waitForNextBlock
+	}
+}
+
+func waitForSpendingRound(restClient client.RestClient, cfg config) {
+	var nodeStatus generatedV2.NodeStatusResponse
+	var err error
+	for {
+		nodeStatus, err = restClient.Status()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "unable to check status : %v", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		if isSpendRound(cfg, nodeStatus.LastRound) {
+			// time to send transactions.
+			return
+		}
+		fmt.Printf("Current round %d, waiting for spending round %d\n", nodeStatus.LastRound, nextSpendRound(cfg, nodeStatus.LastRound))
+	waitForNextBlock:
```

Comment:
Why the repetition inside the labeled block?
Why not just wait for nextPsendRound and continue here?

Reply:
it would repeat the "Status" call:
1. The first time. ( so that we would have a solid knowledge of what is the current round )
2. Every time we had a communication error.

Otherwise, it would perform the WaitForBlock, blocking for exactly one round at a time.